app/models/cohort_lifecycle.rb

```diff
@@ -382,4 +411,178 @@ def self.build_result_hash( obs_data )
 
     result_hash
   end
+
+  def self.get_improving_identifiers( user_id, place_ids )
+    filter = [
+      {
+        bool: {
+          must: [
+            { term: { category: "improving" } },
+            { term: { "taxon.rank": "species" } },
+            { term: { own_observation: false } },
+            { term: { "user.id": user_id } }
+          ]
+        }
+      }
+    ]
+
+    aggregation = {
+      taxon_id: {
+        terms: {
+          field: "taxon.id",
+          size: 10_000,
+          min_doc_count: 3
+        },
+        aggs: {
+          place_ids: {
+            terms: {
+              field: "observation.place_ids",
+              size: 10_000,
+              min_doc_count: 1
+            }
+          }
+        }
+      }
+    }
+
+    response = Identification.elastic_search(
+      size: 0,
+      filters: filter,
+      aggregate: aggregation
+    ).response.aggregations.taxon_id.buckets
+
+    results = response.map do | taxon_bucket |
+      {
+        taxon_id: taxon_bucket["key"],
+        place_ids: taxon_bucket.place_ids.buckets.map {| place_bucket | place_bucket["key"] } & place_ids
+      }
+    end
+    results.map {| a | [a[:taxon_id], a[:place_ids]] }.to_h
+  end
+
+  def self.get_place_obs( oids, place_ids )
```

Comment:
Another minor naming suggestion, I find it much preferable to be explicit with variable names even if they're long. I'd prefer `observation_ids` here over `oids`. Later on you define `observation_ids` in a loop - than could be renamed to `batch_observation_ids` or something like that if `observation_ids` is the method parameter.